CMakeLists.txt

```diff
@@ -0,0 +1,115 @@
+# 1. Requirements
+cmake_minimum_required(VERSION 3.21)
+
+project(
+OpenAssetIO-MediaCreation
+VERSION 1.0.0
+)
+
+set(CMAKE_EXPORT_COMPILE_COMMANDS ON)
+
+# Additional include directories for CMake utils.
+list(APPEND CMAKE_MODULE_PATH ${CMAKE_CURRENT_LIST_DIR}/cmake)
+
+#-----------------------------------------------------------------------
+# Options
+option(OPENASSETIO_MEDIACREATION_ENABLE_TEST "Run test on mediacreation traits" OFF)
+message(STATUS "Test enabled = ${OPENASSETIO_MEDIACREATION_ENABLE_TEST}")
+
+# ABI wrangling only needed for test
+if (CMAKE_CXX_COMPILER_ID STREQUAL "GNU" AND
+    CMAKE_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL 5.0)
+    option(OPENASSETIO_MEDIACREATION_GLIBCXX_USE_CXX11_ABI "For gcc, use the new C++11 library ABI"
+           OFF)
+    message(STATUS "New C++11 ABI for GCC = ${OPENASSETIO_MEDIACREATION_GLIBCXX_USE_CXX11_ABI}")
+endif ()
+
+#-----------------------------------------------------------------------
+# Default install directory
+
+# Default install to a `dist` directory under the build directory, ready
+# for use in tests and for packaging. But don't override if user has
+# explicitly set CMAKE_INSTALL_PREFIX.
+if (CMAKE_INSTALL_PREFIX_INITIALIZED_TO_DEFAULT AND PROJECT_IS_TOP_LEVEL)
+    set(CMAKE_INSTALL_PREFIX "${PROJECT_BINARY_DIR}/dist" CACHE PATH "Installation location" FORCE)
+endif ()
+
+#-----------------------------------------------------------------------
+# OpenAssetIO-MediaCreation generates its traits at configure time, via
+# calling `openassetio-traitgen`.
+# Some might be surprised by this, and expect code generation to be
+# performed in a build step. However, with MediaCreation being a
+# header only library, there is no "build-step" as such, (unless added
+# coincidently, if you are building the test executable, for example.)
+# Furthermore, generating sources at configure time serves to more
+# directly mimic a non-generated project, and lowers risk that changes
+# to the library strucutre later on will be caused issue by the
+# generative nature of the project.
+message("Generating Traits with OpenAssetIO-Traitgen")
+# Mark the traits source as a configure dependency by copying it to the
+# binary dir, and use that file to run traitgen from.
+# This means if the source traits file is changed, configure will rerun.
+configure_file(${CMAKE_CURRENT_LIST_DIR}/traits.yml ${CMAKE_CURRENT_BINARY_DIR}/traits.yml)
+execute_process(COMMAND openassetio-traitgen ${CMAKE_CURRENT_BINARY_DIR}/traits.yml
+                -o ${CMAKE_CURRENT_LIST_DIR} -g cpp)
```

Comment:
[fix] Ideally we would put the generated files under the BINARY_DIR, to avoid polluting your checkout.  

I suggest `${PROJECT_BINARY_DIR}` rather than `${CMAKE_CURRENT_BINARY_DIR}`, since the first is top-level, whereas the second is (potentially) a subdirectory of the build directory, determined by the subdirectory of the `CMakeLists.txt` (in this particular case they're equivalent, so no big deal).

Reply:
The idea I was going for was to make the project as much like the generation wasn't a factor, ie, generate it to where source code would normally be, but fair enough if you think this is the way to go, I can see the point.

https://github.com/OpenAssetIO/OpenAssetIO-MediaCreation/pull/25/commits/c3892a62bc5f12495330cdfe5979fab00dea5fe6